collector/cpu_dragonfly.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !nocpu
+
+package collector
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"unsafe"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+/*
+#cgo LDFLAGS:
+#include <sys/sysctl.h>
+#include <kinfo.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+int
+getCPUTimes(char **cputime) {
+	size_t len;
+
+	// Get number of cpu cores.
+	int mib[2];
+	int ncpu;
+	mib[0] = CTL_HW;
+	mib[1] = HW_NCPU;
+	len = sizeof(ncpu);
+	if (sysctl(mib, 2, &ncpu, &len, NULL, 0)) {
+		return -1;
+	}
+
+	// The bump on each statclock is
+	// ((cur_systimer - prev_systimer) * systimer_freq) >> 32
+	// where
+	// systimer_freq = sysctl kern.cputimer.freq
+	long freq;
+	len = sizeof(freq);
+	if (sysctlbyname("kern.cputimer.freq", &freq, &len, NULL, 0)) {
+		return -1;
+	}
+
+	// Get the cpu times.
+	struct kinfo_cputime cp_t[ncpu];
+	bzero(cp_t, sizeof(struct kinfo_cputime)*ncpu);
+	len = sizeof(cp_t[0])*ncpu;
+	if (sysctlbyname("kern.cputime", &cp_t, &len, NULL, 0)) {
+		return -1;
+	}
+
+	// string needs to hold (5*ncpu)(uint64_t + char)
+	// The char is the space between values.
+	int cputime_size = (sizeof(uint64_t)+sizeof(char))*(5*ncpu);
```

Comment:
sizeof is the size of the binary representation (8), but (if the number is large) the decimal representation you're putting in this array will be longer. 2^64-1 has 20 digits, so I think if you just put in 5*ncpu*21 it'll be enough?

Additionally, instead of malloc+bzero you can use [calloc](http://pubs.opengroup.org/onlinepubs/009695399/functions/calloc.html) below – memory returned from calloc is initialized to 0: `*cputime = calloc(ncpu, 5*21)` should do.

I'm not a C programmer either …